layout/layout.c

```diff
@@ -810,44 +810,38 @@ M0_INTERNAL uint64_t m0_layout_find_by_buffsize(struct m0_layout_domain *dom,
 						struct m0_fid *pver,
 						size_t buffsize)
 {
-	uint64_t lid = M0_DEFAULT_LAYOUT_ID;
-	struct m0_pdclust_attr *pa;
-	struct m0_layout *l;
-	uint64_t hash;
+	struct m0_pdclust_attr *pa = NULL;
+	struct m0_layout *l = NULL;
+	uint64_t hash = 0;
 	int i;
 
 	m0_mutex_lock(&dom->ld_lock);
 	for (i = M0_DEFAULT_LAYOUT_ID; i < m0_lid_to_unit_map_nr; ++i) {
-		/* Current BE max tx size constraints. */
-		if (m0_lid_to_unit_map[i] > 4 * LNET_MAX_PAYLOAD)
-			break;
 		hash = m0_pool_version2layout_id(pver, i);
 		l = m0_layout__list_lookup(dom, hash, true);
 		if (l != NULL) {
 			pa = &m0_layout_to_pdl(l)->pl_attr;
-			if (pa->pa_unit_size * pa->pa_N <= buffsize) {
-				lid = i;
-			} else {
-				m0_ref_put(&l->l_ref);
+			if (pa->pa_unit_size * pa->pa_N >= buffsize ||
+			    /*
+			     * Performance degrades for units bigger than
+			     * 4 * LNET_MAX_PAYLOAD (4MB atm), as tests show.
+			     */
+			    m0_lid_to_unit_map[i] > 4 * LNET_MAX_PAYLOAD)
 				break;
-			}
 			m0_ref_put(&l->l_ref);
 		}
 	}
 
-	hash = m0_pool_version2layout_id(pver, lid);
-	l = m0_layout__list_lookup(dom, hash, true);
 	if (l != NULL) {
-		pa = &m0_layout_to_pdl(l)->pl_attr;
 		M0_LOG(M0_INFO,
 			"Found lid=%d (pver+lid hash=%lx, unit_size=%d, N=%d) "
-			"by buffer size %d.", (int)lid, (unsigned long int)hash,
+			"by buffer size %d.", i, (unsigned long int)hash,
```

Comment:
Here, I think this is a bug:
If the execution breaks from previous loop, the *l* should be a valid pointer with a valid ref. We need to put it.
If the execution ends when the loop finishes iteration ( i == m0_lid_to_unit_map_nr), the *l" is a valid pointer, but the ref count is already dropped.

I'd like to move this into previous loop.
```

	m0_mutex_lock(&dom->ld_lock);
	for (i = M0_DEFAULT_LAYOUT_ID; i < m0_lid_to_unit_map_nr; ++i) {
		hash = m0_pool_version2layout_id(pver, i);
		l = m0_layout__list_lookup(dom, hash, true);
		if (l != NULL) {
			pa = &m0_layout_to_pdl(l)->pl_attr;
			if (pa->pa_unit_size * pa->pa_N >= buffsize ||
			    /*
			     * Performance degrades for units bigger than
			     * 4 * LNET_MAX_PAYLOAD (4MB atm), as tests show.
			     */
			    m0_lid_to_unit_map[i] > 4 * LNET_MAX_PAYLOAD) {
                               m0_ref_put(&l->l_ref);
				break;
                       }
			m0_ref_put(&l->l_ref);
		}
	}
```

Reply:
Good catch! Let me fix that...

Reply:
Fixed, please check now.